ext/magic/extconf.rb

```diff
@@ -46,6 +156,90 @@ def windows?
   RbConfig::CONFIG['target_os'] =~ /mswin|mingw32|windows/
 end
 
+def truffle?
+  ::RUBY_ENGINE == 'truffleruby'
+end
+
+def concat_flags(*args)
+  args.compact.join(" ")
+end
+
+def do_help
+  print(MAGIC_HELP_MESSAGE)
+  exit!(0)
+end
+
+def do_clean
+  root = Pathname(PACKAGE_ROOT_DIR)
+  pwd  = Pathname(Dir.pwd)
+
+  # Skip if this is a development work tree
+  unless (root + '.git').exist?
+    message("Cleaning files only used during build.\n")
+
+    # (root + 'tmp') cannot be removed at this stage because
+    # libmagic.so is yet to be copied to lib.
+
+    # clean the ports build directory
+    Pathname.glob(pwd.join('tmp', '*', 'ports')) do |dir|
+      FileUtils.rm_rf(dir, verbose: true)
+    end
+
+    FileUtils.rm_rf(root + 'ports' + 'archives', verbose: true)
+
+    # Remove everything but share/ directory
+    remove_paths = %w[bin include]
+    remove_paths << 'lib' if config_static?
+
+    Pathname.glob(File.join(root, 'ports', '*', 'libmagic', '*')) do |dir|
+      remove_paths.each do |path|
+        remove_dir = File.join(dir, path)
+        FileUtils.rm_rf(remove_dir, verbose: true)
+      end
+    end
+  end
+
+  exit!(0)
+end
+
+#
+#  main
+#
+do_help if arg_config('--help')
+do_clean if arg_config('--clean')
+
+if config_system_libraries?
+  message "Building ruby-magic using system libraries.\n"
+
+  dir_config('magic')
+else
+  message "Building ruby-magic using packaged libraries.\n"
+
+  static_p = config_static?
+  message "Static linking is #{static_p ? 'enabled' : 'disabled'}.\n"
+  cross_build_p = config_cross_build?
+  message "Cross build is #{cross_build_p ? 'enabled' : 'disabled'}.\n"
+
+  libmagic_recipe = process_recipe('libmagic', LIBMAGIC_TAG, static_p, cross_build_p) do |recipe|
+    recipe.files = [{
```

Comment:
It seems that this might work (not tested):

```ruby
recipe.files = [
	{
		url: "file://#{File.dirname(__FILE__)}/file-#{recipe.version}.tar.gz",
		sha256: LIBIMAGE_SHA256,
	},
	{
		url: "ftp://ftp.astron.com/pub/file/file-#{recipe.version}.tar.gz",
		sha256: LIBIMAGE_SHA256,
	},
	{
		url: "https://ruby-magic.s3.eu-central-1.amazonaws.com/file-#{recipe.version}.tar.gz",
        sha256: LIBIMAGE_SHA256,
	},
]
```

A way to save some S3 traffic. What do you think?

Reply:
@kwilczynski Unfortunately, that doesn't actually work or save traffic. `mini_portile` first checks in https://github.com/flavorjones/mini_portile/blob/4b5455caf2f5f6c65d1b29afeadf014fdd63280a/lib/mini_portile2/mini_portile.rb#L134-L136 whether there are any missing files in the `archives` local directory, and then it tries to download each item (https://github.com/flavorjones/mini_portile/blob/4b5455caf2f5f6c65d1b29afeadf014fdd63280a/lib/mini_portile2/mini_portile.rb#L63-L64) from the list.

At the moment, the S3 source appears to be the only reliable option. The FTP site only handles 10 concurrent users, and fossies.org appears to be rate-limiting by User-Agent.

Reply:
Hi @stanhu,

Oh, I see what you mean by each one from the list. A bit disappointing, I really hoped this would work as multiple sources from which to download... try next should current fails. Then I found this too https://github.com/flavorjones/mini_portile/issues/44.

Thank you for looking into this!

Krzysztof